exe-unit/src/network/inet.rs

```diff
@@ -469,27 +594,37 @@ impl Proxy {
             .left_future());
         }
 
-        log::debug!("[inet] connect to {:?}", desc);
+        self.drop_unused_connections(200).await;
+
+        print_sockets(&network);
+
+        log::debug!("[inet] connect to {desc:?}, using handle: {handle}");
 
-        let (ip, port) = (conv_ip_addr(meta.local.addr)?, meta.local.port);
+        let (ip, port) = (
+            conv_ip_addr(meta.local.addr).map_err(|e| ProxyingError::routeable(conn, e))?,
+            meta.local.port,
+        );
         if let Some(ref filter) = self.filter {
-            filter.validate(meta.protocol, ip, port)?;
+            filter
+                .validate(meta.protocol, ip, port)
+                .map_err(|e| ProxyingError::routeable(conn, e.into()))?;
         }
 
-        let (tx, mut rx) = match meta.protocol {
-            Protocol::Tcp => inet_tcp_proxy(ip, port).await?,
-            Protocol::Udp => inet_udp_proxy(ip, port).await?,
+        let (tx, rx) = match meta.protocol {
+            Protocol::Tcp => inet_tcp_proxy(ip, port).await,
+            Protocol::Udp => inet_udp_proxy(ip, port).await,
             other => return Err(NetError::ProtocolNotSupported(other.to_string()).into()),
-        };
-
-        let conn = Connection { handle, meta };
-        let proxy = self.clone();
+        }
+        .map_err(|e| ProxyingError::routeable(conn, e))?;
 
         let mut state = self.state.write().await;
-        state.remotes.insert(key, tx);
+        let (conn_state, mut rx) = ConnectionState::new(tx, rx);
+        state.remotes.insert(key.clone(), conn_state);
 
         Ok(async move {
             while let Some(bytes) = rx.next().await {
+                proxy.update_seen(&key).await;
+
                 let vec = match bytes {
                     Ok(bytes) => Vec::from_iter(bytes.into_iter()),
```

Comment:
Not sure which is more idiomatic but `collect` is used more often.
```suggestion
                    Ok(bytes) => bytes.into_iter().collect::<Vec<_>>(),
```